### PR TITLE
Inline generative effects in cmm

### PR DESCRIPTION
**Pull request by Gbury**
See the comment in the code.

### CODE REVIEW

**Comment by lthls** [comment]
I think the comment could be improved.
Here is a suggestion:
```
The decision here is whether to consider the binding for inlining or not. It is always correct to consider an effectful expression for inlining, as the environment is going to handle the details of preserving the effects and coeffects ordering (if inlining without reordering is impossible then the expressions will be bound at some safe place instead).
So the decision here is about readability (effectful expressions as arguments to other primitives make it hard to follow the order of evaluation) and locality (for deeply nested expressions, binding the sub-expressions outside can keep them alive for longer than strictly necessary).
The current choice of always inlining pure expressions and expressions with only generative effects is guided by the relatively common case of initialisation of huge static structures (including arrays). Without inlining, all intermediate results would be live for long periods of time and the default register allocator would have trouble dealing with that (it's quadratic in the number of registers live at the same time).
Deep expressions involving arbitrary effects are less common, so inlining for these expressions is controlled by the global [inline_effects_in_cmm] setting.
```
Feel free to cherry-pick any parts of the suggestion that you like.

**Comment by mshinwell** [comment]
The proposed new comment looks fine, so I've pushed that, and will approve.

**Review by mshinwell — APPROVED** [review]
